File: drake_ros_bazel_installed/tools/skylark/ros2/resources/rmw_isolation/test/isolated_listener.py
import os

import rclpy
import rclpy.node
import rclpy.executors
import std_msgs.msg


class IsolatedListener(rclpy.node.Node):

    def __init__(self, uuid):
        super().__init__('isolated_listener')
        self._subscription = self.create_subscription(
            std_msgs.msg.String,
            'uuid',
            self._topic_callback,
            10)
        timeout = self.declare_parameter('timeout', 2.0)
        self._timer = self.create_timer(
            timeout.value, self._timer_callback)
        self._expected_messages_received = 0
        self._uuid = uuid

    def _topic_callback(self, msg):
        assert msg.data == self._uuid, \
            f"I heard '{msg.data}' yet I was expecting '{self._uuid}'!"
        self._expected_messages_received += 1

    def _timer_callback(self):
        assert self._expected_messages_received > 0, \
            f"I did not hear '{elf._uuid}' even once!"
        rclpy.shutdown()


def main():
    if 'TEST_TMPDIR' in os.environ:
        from rmw_isolation import isolate_rmw_by_path
        isolate_rmw_by_path(os.environ['TEST_TMPDIR'])

    rclpy.init()
    uuid = os.environ.get('TEST_TMPDIR', 'none')
    try:
        executor = rclpy.executors.SingleThreadedExecutor()
        rclpy.spin(IsolatedListener(uuid), executor)
    finally:
        # NOTE(hidmic): try_shutdown raises AttributeError
        # Need https://github.com/ros2/rclpy/pull/812
        pass  # rclpy.try_shutdown()


if __name__ == '__main__':
    main()

File: drake_ros_bazel_installed/tools/skylark/ros2/resources/rmw_isolation/test/isolated_listener.cc
#include <chrono>
#include <memory>

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

#include "rmw_isolation/rmw_isolation.h"

using std::placeholders::_1;

class IsolatedListener : public rclcpp::Node {
 public:
  IsolatedListener(const std::string& uuid)
      : rclcpp::Node("isolated_listener"), uuid_(uuid) {
    double timeout = this->declare_parameter("timeout", 2.0);
    subscription_ = this->create_subscription<std_msgs::msg::String>(
        "uuid", 10, std::bind(&IsolatedListener::topic_callback, this, _1));
    timer_ = this->create_wall_timer(
        std::chrono::duration<double>(timeout),
        std::bind(&IsolatedListener::timer_callback, this));
  }

 private:
  void topic_callback(const std_msgs::msg::String::SharedPtr msg) {
    if (msg->data != uuid_) {
      throw std::runtime_error("I heard '" + msg->data + "' yet " +
                               "I was expecting '" + uuid_ + "'!");
    }
    ++expected_messages_count_;
  }

  void timer_callback() const {
    if (0u == expected_messages_count_) {
      throw std::runtime_error("I did not hear '" + uuid_ + "' even once!");
    }
    rclcpp::shutdown();
  }

  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
  rclcpp::TimerBase::SharedPtr timer_;
  size_t expected_messages_count_{0u};
  std::string uuid_;
};

int main(int argc, char* argv[]) {
  const char* TEST_TMPDIR = std::getenv("TEST_TMPDIR");
  if (TEST_TMPDIR != nullptr) {
    ros2::isolate_rmw_by_path(argv[0], TEST_TMPDIR);
  }
  rclcpp::init(argc, argv);
  std::string uuid{TEST_TMPDIR != nullptr ? TEST_TMPDIR : "none"};
  rclcpp::spin(std::make_shared<IsolatedListener>(uuid));
  rclcpp::shutdown();
  return 0;
}
